designs/multi-cluster.md

```diff
@@ -113,104 +131,172 @@ type Cluster interface {
+the engaged cluster (e.g. because the events do not supply information about the cluster in
+object annotations) that it has been started for.
 
 The `ctrl.Manager` _SHOULD_ be extended by a `cluster.Cluster` getter:
+
 ```golang
 // pkg/manager
 type Manager interface {
     // ...
     GetCluster(ctx context.Context, clusterName string) (cluster.Cluster, error)
 }
 ```
+
 The embedded `cluster.Cluster` corresponds to `GetCluster(ctx, "")`. We call the
 clusters with non-empty name "provider clusters" or "enganged clusters", while
 the embedded cluster of the manager is called the "default cluster" or "hub 
 cluster".
 
-The `reconcile.Request` _SHOULD_ be extended by an optional `ClusterName` field:
+
+### Multi-Cluster-Compatible Reconcilers
+
+Reconcilers can be made multi-cluster-compatible by changing client and cache
+accessing code from directly accessing `mgr.GetClient()` and `mgr.GetCache()` to
+going through `mgr.GetCluster(clusterName).GetClient()` and
+`mgr.GetCluster(clusterName).GetCache()`. `clusterName` needs to be extracted from
+the BYO `request` type (e.g. a `clusterName` field in the type itself).
+
+A typical snippet at the beginning of a reconciler to fetch the client could look like this:
+
 ```golang
-// pkg/reconile
-type Request struct {
-    ClusterName string
-    types.NamespacedName
+cl, err := mgr.GetCluster(ctx, req.ClusterName)
+if err != nil {
+	return reconcile.Result{}, err
 }
+client := cl.GetClient()
 ```
 
-With these changes, the behaviour of controller-runtime without a set cluster
-provider will be unchanged.
+Due to the BYO `request` type, controllers need to be built like this:
 
-### Multi-Cluster-Compatible Reconcilers
+```golang
+builder.TypedControllerManagedBy[ClusterRequest](mgr).
+	Named("multi-cluster-controller").
+	Watches(&corev1.Pod{}, &ClusterRequestEventHandler{}).
+	Complete(reconciler)
+```
 
-Reconcilers can be made multi-cluster-compatible by changing client and cache
-accessing code from directly accessing `mgr.GetClient()` and `mgr.GetCache()` to
-going through `mgr.GetCluster(req.ClusterName).GetClient()` and
-`mgr.GetCluster(req.ClusterName).GetCache()`.
+With `ClusterRequest` and `ClusterRequestEventHandler` being BYO types. `reconciler`
+can be e.g. of type `reconcile.TypedFunc[ClusterRequest]`.
+
+`ClusterRequest` will likely often look like this:
 
-When building a controller like
 ```golang
-builder.NewControllerManagedBy(mgr).
-    For(&appsv1.ReplicaSet{}).
-	Owns(&v1.Pod{}).
-    Complete(reconciler)
+type ClusterRequest struct {
+	reconcile.Request
+	ClusterName string
+}
 ```
-with the described change to use `GetCluster(ctx, req.ClusterName)` will automatically 
-act as *uniform multi-cluster controller*. It will reconcile resources from cluster `X`
+
+Controllers that use `For` or `Owns` cannot be converted to multi-cluster controllers
+without changing to `Watches` as the BYO `request` type cannot be used with them:
```

Comment:
that's unfortunate 🤔

Reply:
Removed the duplication and explicitly made a statement about `Owns` not being usable without a re-implementation of the event handler.